fdbserver/BulkDumpUtil.actor.cpp

```diff
@@ -0,0 +1,229 @@
+}
+
+std::string generateBulkDumpDataFileName(Version version) {
+	return std::to_string(version) + "-data.sst";
+}
+
+std::string generateBulkDumpManifestFileName(Version version) {
+	return std::to_string(version) + "-manifest.txt";
+}
+
+std::string generateBulkDumpByteSampleFileName(Version version) {
+	return std::to_string(version) + "-sample.sst";
+}
+
+BulkDumpManifest dumpDataFileToLocalDirectory(const std::map<Key, Value>& sortedKVS,
+                                              const std::string& rootFolder,
+                                              const std::string& relativeFolder,
+                                              Version dumpVersion,
+                                              const KeyRange& dumpRange,
+                                              int64_t dumpBytes) {
+	const std::string dumpFolder = abspath(joinPath(rootFolder, relativeFolder));
+	platform::eraseDirectoryRecursive(dumpFolder);
+	if (!platform::createDirectory(dumpFolder)) {
+		throw retry();
+	}
+
+	std::string dataFilePath = "";
+	std::string byteSampleFilePath = "";
+	if (sortedKVS.size() > 0) {
+		std::string dataFileName = generateBulkDumpDataFileName(dumpVersion);
+		dataFilePath = abspath(joinPath(dumpFolder, dataFileName));
+		ASSERT(!fileExists(dataFilePath));
+		std::string byteSampleFileName = generateBulkDumpByteSampleFileName(dumpVersion);
+		byteSampleFilePath = abspath(joinPath(dumpFolder, byteSampleFileName));
+		ASSERT(!fileExists(byteSampleFilePath));
+
+		std::unique_ptr<IRocksDBSstFileWriter> sstWriterData = newRocksDBSstFileWriter();
+		std::unique_ptr<IRocksDBSstFileWriter> sstWriterByteSample = newRocksDBSstFileWriter();
+		sstWriterData->open(dataFilePath);
+		sstWriterByteSample->open(byteSampleFilePath);
+		bool anySampled = false;
+		for (const auto& [key, value] : sortedKVS) {
+			sstWriterData->write(key, value); // assuming sorted
+			ByteSampleInfo sampleInfo = isKeyValueInSample(KeyValueRef(key, value));
+			if (sampleInfo.inSample) {
+				sstWriterByteSample->write(key, value);
+				anySampled = true;
+			}
+		}
+		ASSERT(sstWriterData->finish());
+		if (anySampled) {
+			ASSERT(sstWriterByteSample->finish());
+		} else {
+			ASSERT(deleteFile(byteSampleFilePath));
+		}
+	}
+
+	std::string manifestFileName = generateBulkDumpManifestFileName(dumpVersion);
+	std::string manifestFilePath = abspath(joinPath(dumpFolder, manifestFileName));
+	ASSERT(!fileExists(manifestFilePath));
+
+	BulkDumpManifest manifest(
+	    dataFilePath, manifestFilePath, byteSampleFilePath, dumpRange.begin, dumpRange.end, dumpVersion, "", dumpBytes);
+	writeFile(manifestFilePath, manifest.toString());
+
+	return manifest;
+}
+
+void bulkDumpFileCopy(std::string fromFile, std::string toFile, size_t fileBytesMax) {
+	std::string content = readFileBytes(fromFile, fileBytesMax);
+	writeFile(toFile, content);
+	return;
+}
+
+ACTOR Future<Void> bulkDumpTransportCP_impl(std::string fromRoot,
```

Comment:
Is this copy necessary? We've dumped to a local dir and then we'd copy elsewhere? Is that what is?

Reply:
This is just for testing without S3. Copy file to a different folder.